File: 0005-longest-palindromic-substring/main.js
// Algorithm 3, based off of Leetcode (editorial)[https://leetcode.com/problems/longest-palindromic-substring/solutions/3598120/longest-palindromic-substring/?page=3]

var longestPalindrome = function(s) {
  var expand = function(left, right) {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      left--;
      right++;
    }
    return s.slice((left + 1), right);
  }

  let ans = "";

  for (let i = 0; i < s.length; i++) {
    let odd = expand(i, i);
    if (odd.length > ans.length) {
      ans = odd;
    }
    let even = expand(i, i + 1)
    if (even.length > ans.length) {
      ans = even;
    }
  }

  return ans;
};

// let s = "abbcdef";
// let r = longestPalindrome(s);
// let e = "bb";
// console.log(e);
// console.log(r === e);


// s = "racecar";
// r = longestPalindrome(s);
// e = "racecar";
// console.log(e);
// console.log(r === e);

File: 0005-longest-palindromic-substring/main.go
package main
import ("fmt")

// Algorithm 3, based off of Leetcode (editorial)[https://leetcode.com/problems/longest-palindromic-substring/solutions/3598120/longest-palindromic-substring/?page=3]
// ported from Javascript
func longestPalindrome(s string) string {
	expand := func(left int, right int) string {
		for left >= 0 && right < len(s) && s[left] == s[right] {
			left--;
			right++;
		}
		sstr := s[(left +1):right]
		return (sstr)
	}

	ans := ""

	for i := 0; i < len(s); i++ {
		odd := expand(i, i)
		if len(odd) > len(ans) {
			ans = odd
		}

		even := expand(i, i + 1)
		if len(even) > len(ans) {
			ans = even
		}
	}

	return ans
}


func main() {
	s := "abbcdef"
	r := longestPalindrome(s)
	e := "bb"
	fmt.Println(e)
	fmt.Println(r == e)

	s = "racecar"
	r = longestPalindrome(s)
	e = "racecar"
	fmt.Println(e)
	fmt.Println(r == e)
}